cmd/vicadmin/server.go

```diff
@@ -112,39 +127,208 @@ func (s *server) listenPort() int {
+		// log out, disregard errors
+		usersession.Client.Logout(context.Background())
+
+		// save user config locally
+		usersess := s.uss.Add(req.FormValue("username"), &userconfig)
+
+		// create a token to save as an encrypted & signed cookie
+		websession, err := s.uss.cookies.Get(req, sessionCookieKey)
+		if websession == nil {
+			log.Errorf("Web session object could not be created due to error %s", err)
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		timeNow, err := usersess.created.MarshalText()
+		if err != nil {
+			log.Errorf("Failed to unmarshal time object %+v into text due to error: %s", usersess.created, err)
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		websession.Values[sessionCreationTimeKey] = string(timeNow)
+		websession.Values[sessionUsernameKey] = req.FormValue("username")
+		if err := websession.Save(req, res); err != nil {
+			log.Errorf("\"%s\" occurred while trying to save session to browser", err.Error())
+			http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+			return
+		}
+
+		// redirect to dashboard
+		http.Redirect(res, req, "/", http.StatusTemporaryRedirect)
+		return
+	}
+
+	// Render login page (shows up on non-POST requests):
+	sess, err := client(&rootConfig)
+	if err != nil {
+		log.Errorf("Could not render login page due to vSphere connection error: %s", err.Error())
+		http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+		return
+	}
+	v := vicadmin.NewValidator(ctx, &vchConfig, sess)
+	tmpl, err := template.ParseFiles("auth.html")
+	err = tmpl.ExecuteTemplate(res, "auth.html", v)
+	if err != nil {
+		log.Errorf("Error parsing template: %s", err)
+		http.Error(res, genericErrorMessage, http.StatusInternalServerError)
+		return
+	}
+}
+
 func (s *server) serve() error {
 	defer trace.End(trace.Begin(""))
 
 	s.mux = http.NewServeMux()
 
+	// s.mux.HandleFunc bypasses authentication
+	s.mux.HandleFunc(loginPagePath, s.loginPage)
+
 	// tar of appliance system logs
-	s.mux.HandleFunc("/logs.tar.gz", s.tarDefaultLogs)
-	s.mux.HandleFunc("/logs.zip", s.zipDefaultLogs)
+	s.Authenticated("/logs.tar.gz", s.tarDefaultLogs)
+	s.Authenticated("/logs.zip", s.zipDefaultLogs)
 
 	// tar of appliance system logs + container logs
-	s.mux.HandleFunc("/container-logs.tar.gz", s.tarContainerLogs)
-	s.mux.HandleFunc("/container-logs.zip", s.zipContainerLogs)
+	s.Authenticated("/container-logs.tar.gz", s.tarContainerLogs)
+	s.Authenticated("/container-logs.zip", s.zipContainerLogs)
 
-	s.mux.Handle("/css/", http.StripPrefix("/css/", http.FileServer(http.Dir("css/"))))
-	s.mux.Handle("/images/", http.StripPrefix("/images/", http.FileServer(http.Dir("images/"))))
-	s.mux.Handle("/fonts/", http.StripPrefix("/fonts/", http.FileServer(http.Dir("fonts/"))))
+	// these assets bypass authentication & are world-readable
+	s.Handle("/css/", http.StripPrefix("/css/", http.FileServer(http.Dir("css/"))))
```

Comment:
There is some code that disallows `..` in paths for example (in the FileServer), but only as a basic precaution - the comments on that package say you should sanitize the paths as the path is essentially untrusted user input